pkgs/development/python-modules/elasticsearch/default.nix

```diff
@@ -1,29 +1,78 @@
-{ buildPythonPackage
-, fetchPypi
-, urllib3, requests
-, nosexcover, mock
-, lib
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+
+# docs
+, sphinx-rtd-theme
+, sphinxHook
+
+# propagates
+, elastic-transport
+
+# extras
+, aiohttp
+, requests
+
+# tests
+, numpy
+, pandas
+, pytest-asyncio
+, pytestCheckHook
+, python-dateutil
+, pyyaml
 }:
 
 buildPythonPackage (rec {
   pname = "elasticsearch";
-  version = "7.16.3";
+  version = "8.4.2";
+  format = "setuptools";
+
+  src = fetchFromGitHub {
+    owner = "elastic";
+    repo = "elasticsearch-py";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-OjgfqEcKQH8ESvkcr2Ue2Q7KW2ykfrZccxm8kxzOClU=";
+  };
+
+  postPatch = ''
+    sed -i '/addopts/d' setup.cfg
+  '';
+
+  outputs = [
+    "out"
+    "doc"
+  ];
+
+  nativeBuildInputs = [
+    sphinx-rtd-theme
+    sphinxHook
+  ];
+
+  sphinxRoot = "docs/sphinx";
+
+  propagatedBuildInputs = [
+    elastic-transport
+  ];
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "8adf8bc351ed55df7296be1009d38a1c999c0abc7d8700fa88533f1ad6087c5e";
+  passthru.optional-dependencies = {
+    async = [ aiohttp ];
+    requests = [ requests ];
   };
 
-  # Check is disabled because running them destroy the content of the local cluster!
-  # https://github.com/elasticsearch/elasticsearch-py/tree/master/test_elasticsearch
-  doCheck = false;
-  propagatedBuildInputs = [ urllib3 requests ];
-  buildInputs = [ nosexcover mock ];
+  checkInputs = [
+    numpy
+    pandas
+    pytest-asyncio
+    pytestCheckHook
+    python-dateutil
+    pyyaml
+  ] ++ lib.flatten (builtins.attrValues passthru.optional-dependencies);
```

Comment:
```suggestion
  ] ++ lib.flatten (lib.attrValues passthru.optional-dependencies);
```

Reply:
Why do we prefer lib over builtins?

Reply:
There is some fallback code in lib https://github.com/NixOS/nixpkgs/blob/master/lib/attrsets.nix#L194

Reply:
Thanks.